x-pack/plugins/enterprise_search/public/applications/app_search/components/search_ui/components/search_ui_form.tsx

```diff
@@ -0,0 +1,205 @@
+  'xpack.enterpriseSearch.appSearch.engine.searchUI.urlFieldPlaceholder',
+  { defaultMessage: 'Select a URL field' }
+);
+const GENERATE_PREVIEW_BUTTON_LABEL = i18n.translate(
+  'xpack.enterpriseSearch.appSearch.engine.searchUI.generatePreviewButtonLabel',
+  { defaultMessage: 'Generate a Preview' }
+);
+
+interface Option<T> extends Omit<EuiComboBoxOptionOption<T>, 'value'> {
+  value: string;
+}
+
+export const SearchUIForm: React.FC = () => {
+  const {
+    validFields,
+    validSortFields,
+    validFacetFields,
+    titleField,
+    urlField,
+    facetFields,
+    sortFields,
+  } = useValues(SearchUILogic);
+  const { engineName } = useValues(EngineLogic);
+  const {
+    onActiveFieldChange,
+    onFacetFieldsChange,
+    onSortFieldsChange,
+    onTitleFieldChange,
+    onUrlFieldChange,
+  } = useActions(SearchUILogic);
+
+  const previewHref = generatePreviwUrl(
+    {
+      fromKibana: 'true',
+      titleField,
+      urlField,
+      facets: facetFields,
+      sortFields,
+    },
+    engineName
+  );
+
+  const formatSelectOption = (fieldName: string) => {
+    return { text: fieldName, value: fieldName };
+  };
+  const formatMultiOptions = (fieldNames: string[]) =>
+    fieldNames.map((fieldName) => ({ label: fieldName, text: fieldName, value: fieldName }));
+  const formatMultiOptionsWithEmptyOption = (fieldNames: string[]) => [
+    { label: '', text: '', value: '' },
+    ...formatMultiOptions(fieldNames),
+  ];
+
+  const optionFields = formatMultiOptionsWithEmptyOption(validFields);
+  const sortOptionFields = formatMultiOptions(validSortFields);
+  const facetOptionFields = formatMultiOptions(validFacetFields);
+  const selectedTitleOption = formatSelectOption(titleField);
+  const selectedURLOption = formatSelectOption(urlField);
+  const selectedSortOptions = formatMultiOptions(sortFields);
+  const selectedFacetOptions = formatMultiOptions(facetFields);
+
+  return (
+    <div>
+      <EuiForm>
+        <EuiFormRow label={TITLE_FIELD_LABEL} helpText={TITLE_FIELD_HELP_TEXT} fullWidth>
+          <EuiSelect
+            name="title-field"
+            options={optionFields}
+            value={selectedTitleOption && selectedTitleOption.value}
+            onChange={(e) => onTitleFieldChange(e.target.value)}
+            fullWidth
+            placeholder={TITLE_FIELD_PLACEHOLDER}
+            onFocus={() => onActiveFieldChange(ActiveField.Title)}
+            onBlur={() => onActiveFieldChange(ActiveField.None)}
+            hasNoInitialSelection
+            data-test-subj="selectTitle"
+          />
+        </EuiFormRow>
+        <EuiFormRow label={FILTER_FIELD_LABEL} helpText={FILTER_FIELD_HELP_TEXT} fullWidth>
+          <EuiComboBox
+            options={facetOptionFields}
+            selectedOptions={selectedFacetOptions}
+            onChange={(newValues) =>
+              onFacetFieldsChange(newValues.map((field) => (field as Option<string>).value))
+            }
```

Comment:
I'm not a super huge fan of the extra interface/generic needed here - any chance you'd be open to a more concise way of getting types working?

```suggestion
          onChange={(newValues) => onFacetFieldsChange(newValues.map((field) => field.value!))}
```

Reply:
That's a gem. I need to remember to use that more often.

Reply:
4639e8d92431bfdaccf1c2f89e8f7ff70964aa0d